In [0]:
import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [0]:
#import data from google drive
from google.colab import drive

In [3]:
#mount the drive
drive.mount("/content/drive")

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [4]:
!ls -all './drive/My Drive/SVHN_neural_network_dataset'
 

total 480122
-r-------- 1 root root 491644096 Oct  3 12:20 SVHN_single_grey1.h5


In [5]:
#import os to list the files available in the folder.
#the folder below will list the directories available in the given path
import os
filePath = './drive/My Drive/SVHN_neural_network_dataset'
os.listdir(filePath)

['SVHN_single_grey1.h5']

In [0]:
#Read the .h5 file into google colab object
dataFile = filePath+'/'+'SVHN_single_grey1.h5'

In [0]:
#This import is required to read hdf5 files into python.
#More info can be found in http://docs.h5py.org/en/latest/quick.html
import h5py

In [8]:
#File content
dataFile

'./drive/My Drive/SVHN_neural_network_dataset/SVHN_single_grey1.h5'

In [0]:
#Read the content of the file into another variable. The file will be read into the variable
f = h5py.File(dataFile, 'r')

In [10]:
print("Keys: %s" % f.keys())
a_group_key = list(f.keys())[0]

Keys: KeysView(<HDF5 file "SVHN_single_grey1.h5" (mode r)>)


In [0]:
data = list(f[a_group_key])

In [12]:
#List the keys present in the dataset
list(f.keys())

['X_test', 'X_train', 'X_val', 'y_test', 'y_train', 'y_val']

In [13]:
#print the keys in the dataset
for key in f.keys():
  print("Key value ",key)
  

Key value  X_test
Key value  X_train
Key value  X_val
Key value  y_test
Key value  y_train
Key value  y_val


In [14]:
#Get the dataset for each of the keys present in the file
for i, key in enumerate(f.keys()):
  print('key value is ',i+1, key)
  X_train = f['X_train'].value
  if(key == 'X_test'):
    X_test = f['X_test'].value
  if(key == 'X_val'):
    X_val = f['X_val'].value
  if(key == 'y_test'):
    y_test = f['y_test'].value
  if(key == 'y_train'):
    y_train = f['y_train'].value
  if(key == 'y_val'):
    y_val = f['y_val'].value
    
    

key value is  1 X_test
key value is  2 X_train
key value is  3 X_val
key value is  4 y_test
key value is  5 y_train
key value is  6 y_val


In [15]:
print("Size of various datasets extracted are ")
print("X_train shape is ", X_train.shape,", data set Dtype is ", X_train.dtype, " ,type ", type(X_train))
print("X_test shape is ", X_test.shape,", data set Dtype is ", X_test.dtype, " ,type ", type(X_test))
print("X_val shape is ", X_val.shape,", data set Dtype is ", X_val.dtype, " ,type ", type(X_val))
print("y_train shape is ", y_train.shape,", data set Dtype is ", y_train.dtype, " ,type ", type(y_train))
print("y_test shape is ", y_test.shape,", data set Dtype is ", y_test.dtype, " ,type ", type(y_test))
print("y_val shape is ", y_val.shape,", data set Dtype is ", y_val.dtype, " ,type ", type(y_val))

Size of various datasets extracted are 
X_train shape is  (42000, 32, 32) , data set Dtype is  float32  ,type  <class 'numpy.ndarray'>
X_test shape is  (18000, 32, 32) , data set Dtype is  float32  ,type  <class 'numpy.ndarray'>
X_val shape is  (60000, 32, 32) , data set Dtype is  float32  ,type  <class 'numpy.ndarray'>
y_train shape is  (42000,) , data set Dtype is  uint8  ,type  <class 'numpy.ndarray'>
y_test shape is  (18000,) , data set Dtype is  uint8  ,type  <class 'numpy.ndarray'>
y_val shape is  (60000,) , data set Dtype is  uint8  ,type  <class 'numpy.ndarray'>


In [0]:
import sklearn

In [0]:
from sklearn import neighbors

In [0]:
from sklearn.neighbors import KNeighborsClassifier

In [0]:
#Generally the optimum way to calculate the K value is below
#Optimum K =  sqrt(N)/2 where N is the number of samples.
#Here we have 42000 samples and hence K = sqrt(42000)/2 = 103
knn = KNeighborsClassifier(n_neighbors=103, weights='uniform', algorithm='auto', n_jobs = -1)

In [0]:
#The current dataset shape is three dimensional which will not cthompile when passed to ML algorithms. KNN always
#expects a two dimensional data.
#So, we will need to reshape data.
#We will need to flatten the three dimensional data into two dimensional data. The number of samples shall be
#maintained whereas The nx and ny components can be multiplied and kept as the second dimension.
#So, (number_of_samples, nx, ny ) will be flattened out as (number_of_samples, nx*ny)

In [0]:
def reshapedata(X):
  number_of_samples, nx, ny = X.shape
  X = X.reshape((number_of_samples, nx*ny))
  print(X.shape)
  return X

In [21]:
X_train = reshapedata(X_train)
print("Shape of X_train after reshaping ",X_train.shape)
X_test = reshapedata(X_test)
print("Shape of X_test after reshaping ",X_test.shape)
X_val = reshapedata(X_val)
print("Shape of X_val after reshaping ", X_val.shape)

(42000, 1024)
Shape of X_train after reshaping  (42000, 1024)
(18000, 1024)
Shape of X_test after reshaping  (18000, 1024)
(60000, 1024)
Shape of X_val after reshaping  (60000, 1024)


In [22]:
#Fit KNN with training data after reshaping the data.
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=-1, n_neighbors=103, p=2,
           weights='uniform')

In [0]:
y_predict = knn.predict(X_test)

In [0]:
from sklearn.metrics import accuracy_score, classification_report, f1_score, confusion_matrix

In [25]:
print("Accuracy score %5.2f " %(accuracy_score(y_true=y_test, y_pred=y_predict)))
print("F1 score is %5.2f" %(f1_score(y_true=y_test, y_pred=y_predict, average='weighted')))
print("Classification report")
print(classification_report(y_test, y_predict))
print("Confusion matrix ")
print(confusion_matrix(y_test, y_predict))

Accuracy score  0.53 
F1 score is  0.53
Classification report
             precision    recall  f1-score   support

          0       0.43      0.71      0.53      1814
          1       0.44      0.73      0.55      1828
          2       0.66      0.50      0.57      1803
          3       0.51      0.43      0.47      1719
          4       0.65      0.64      0.65      1812
          5       0.55      0.41      0.47      1768
          6       0.54      0.40      0.46      1832
          7       0.67      0.63      0.65      1808
          8       0.51      0.39      0.44      1812
          9       0.54      0.47      0.50      1804

avg / total       0.55      0.53      0.53     18000

Confusion matrix 
[[1287   82   25   43   52   29   90   39   51  116]
 [ 118 1328   53   62   68   31   36   66   30   36]
 [ 105  254  906   71   61   53   31  161   60  101]
 [ 131  282   90  740   51  149   31   78   95   72]
 [ 134  251   42   56 1167   21   51   17   26   47]
 [ 179  176   47

In [0]:
#Let us go with neural networks to run the test cases again
#START -----  NEURAL NETWORKS -----#

In [0]:
#Let us create a model
model = tf.keras.models.Sequential()

In [0]:
#Let us normalize the data
X_train = X_train/1024
X_test = X_test/1024
X_val = X_val/1024

In [0]:
#Let us convert Y values to categorical
y_train = tf.keras.utils.to_categorical(y_train, num_classes=10)
y_test = tf.keras.utils.to_categorical(y_test, num_classes=10)
y_val = tf.keras.utils.to_categorical(y_val, num_classes=10)

In [30]:
#Add batch normalization
model.add(tf.keras.layers.BatchNormalization())
#Add a dense layer with 200 neurons
tf.keras.layers.Dense(units=200, activation='relu', input_shape=(1024,))

In [0]:
#Let us add few more layers with units=100, 50, 20 and output layer with unit as 10
model.add(tf.keras.layers.Dense(units=100, activation='relu'))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Dense(units=50, activation='relu'))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Dense(units=20, activation='relu'))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Dense(units=10, activation='softmax'))

In [0]:
#Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [33]:
#Fit the data and execute the model
model.fit(x=X_train, y=y_train, batch_size=64, epochs=30, validation_data=(X_val, y_val))

Train on 42000 samples, validate on 60000 samples
Epoch 1/30
42000/42000 [==============================] - 9s 221us/step - loss: 2.3034 - acc: 0.1002 - val_loss: 2.3025 - val_acc: 0.1000
Epoch 2/30
42000/42000 [==============================] - 8s 181us/step - loss: 2.3030 - acc: 0.0999 - val_loss: 2.3026 - val_acc: 0.1000
Epoch 3/30
42000/42000 [==============================] - 8s 179us/step - loss: 2.0247 - acc: 0.2433 - val_loss: 1.5621 - val_acc: 0.4619
Epoch 4/30
42000/42000 [==============================] - 7s 177us/step - loss: 1.3611 - acc: 0.5468 - val_loss: 1.1854 - val_acc: 0.6212
Epoch 5/30
42000/42000 [==============================] - 8s 179us/step - loss: 1.1283 - acc: 0.6387 - val_loss: 1.0314 - val_acc: 0.6739
Epoch 6/30
42000/42000 [==============================] - 7s 178us/step - loss: 0.9859 - acc: 0.6926 - val_loss: 0.9329 - val_acc: 0.7098
Epoch 7/30
42000/42000 [==============================] - 8s 179us/step - loss: 0.8869 - acc: 0.7256 - val_loss: 0.8525 - 

In [0]:
y_pred = model.predict(X_test)

In [35]:
model.evaluate(X_test, y_test)

18000/18000 [==============================] - 1s 50us/step


[0.6644137847688463, 0.8145]

Observation:
1. Traditional ML - KNN algorithm:
  1. The prediction took a lot of time to predict the test values which were fed into the model
  2. The accuracy score was hovering around 53% which was very low compared with the neural network model.
2. Neural network model - using batch normalization, RELU as activation, ADAM as optimizer
  1. The model took less than 4 minutes to train the dataset and the prediction, evaluation was even faster, unlike KNN which took more than 10 minutes to completed
  2. Inspite of being faster, there was no issues with accuracy and other metric parameters.
  3. The accuracy in the train data set was 87.3%, validation set was 85.9%and the test set gave an accuracy of 81.45% which is exceedingly high compared with the traditional ML models.

In [0]:
#Picked from Internet - custom metrics. ----TO CHECK THIS LATER -----
#Let us check the version of keras used.
tf.keras.__version__

'2.1.6-tf'

In [0]:
#Since Keras 2.0, metrics like precision, recall, f1 score have been removed from the libraries. Hence we would
#be required to write custom metric functions for calculating them
#The below code is copied from https://github.com/keras-team/keras/issues/5400

from keras import backend as K

Using TensorFlow backend.


In [0]:
def precision(y_true, y_pred):
    """Precision metric.

    Only computes a batch-wise average of precision.

    Computes the precision, a metric for multi-label classification of
    how many selected items are relevant.
    """
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
    precision = true_positives / (predicted_positives + K.epsilon())
    return precision

def recall(y_true, y_pred):
    """Recall metric.

    Only computes a batch-wise average of recall.

    Computes the recall, a metric for multi-label classification of
    how many relevant items are selected.
    """
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
    recall = true_positives / (possible_positives + K.epsilon())
    return recall


def f1(y_true, y_pred):
    def recall(y_true, y_pred):
        """Recall metric.

        Only computes a batch-wise average of recall.

        Computes the recall, a metric for multi-label classification of
        how many relevant items are selected.
        """
        true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
        possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
        recall = true_positives / (possible_positives + K.epsilon())
        return recall

    def precision(y_true, y_pred):
        """Precision metric.

        Only computes a batch-wise average of precision.

        Computes the precision, a metric for multi-label classification of
        how many selected items are relevant.
        """
        true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
        predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
        precision = true_positives / (predicted_positives + K.epsilon())
        return precision
    precision = precision(y_true, y_pred)
    recall = recall(y_true, y_pred)
    return 2*((precision*recall)/(precision+recall+K.epsilon()))